cdtb/rstfile.py

```diff
@@ -1,24 +1,34 @@
 import os
-from xxhash import xxh64_intdigest
+from xxhash import xxh3_64_intdigest, xxh64_intdigest
 from base64 import b64encode
 from .tools import BinaryParser
 from .hashes import HashFile, default_hash_dir
 
 
-def key_to_hash(key, bits=40):
+def key_to_hash(key, bits=40, rsthash_version=1415):
```

Comment:
We could use the following to limit the use of `bits`/`rsthash_version` parameters to just few places.
```py
def get_key_hasher(bits=40, rsthash_version=1415):
    if rsthash_version >= 1415:
        func = xxh3_64_intdigest
    else:
        func = xxh64_intdigest
    def key_hasher(key):
        func(key) & ((1 << bits) - 1)
    return key_hasher
```

The `key_hasher` could be retrieved as soon as possible. And the API will not have to change if we have to introduce new parameters or whatever.

For instance, for `RstFile` constructor
```py
class RstFile:
    def __init__(self, path_or_f=None, key_hasher=None):
        ...
        if key_hasher is None:
            key_hasher = get_key_hasher()
        self.key_hasher = key_hasher
```
Then use `h = self.key_hasher(key)`

Reply:
hmm how would that work though? Ideally the `RstFile` class is the only place that knows what game version maps to what key function and what rst version maps to what bits.
Your suggestion assumes that the calling code is responsible for creating a key_hasher and needs to know what `bits` and `rsthash_version` to provide.

That said, we could construct something like that internally in `RstFile` to reduce `key_to_hash` invocations on every dict(-like) access.

Reply:
> Ideally the RstFile class is the only place that knows what game version maps to what key function and what rst version maps to what bits.

The RstFile class should be generic and should not have to know about Riot shenanigans. It should be a simple container for translation strings (and their hashes/identifiers).

But more importantly: only the top-level code (updater, etc.) know about patch versions. And it's better if those concepts stay there (at high level, in the updater and its friends).
So yes, I totally assume that the calling code *must* know the "hash format" used for its file. RstFile doesn't need to know why or how hash formats are different, it just needs to know how to hash (hence the generic function parameter).
Having this intermediate type avoids to tie low-level and high-level stuff; which is usually a good idea.

We do provide a helper (`get_key_hasher()`) to help get a correct hasher. It simple to use and we can easily extend it as needed.

Reply:
I still don't understand how you would construct such a key_hasher from that outside before instantiating an `RstFile`. How would the calling code know the correct value for `bits`, when that information is necessarily tied to each individual `RstFile`?

Reply:
Hum... You're right. I forgot the bit size was read from RST header and updated after its initialization. Sorry for that.
I assumed that if there was a version, Riot would have incremented it when changing the hash format... but it's Riot. :(

IMHO defining a separate `get_key_hasher()` would be a good idea anyway.
But as you said, it cannot be computed until final bit size value is known.